### PR TITLE
tests: Run 'screen -ls' in a busy-waiting loop

### PR DESCRIPTION
**Pull request by gc-plp**
Signed-off-by: Gabriel Ionescu <gbi@amazon.com>

## Reason for This PR

Occasionally tests would fail due to the fact that the output of 'screen -ls' did not contain the screen session name.

Due to this issue, re.search() would return 'None', which would throw a generic exception.

## Description of Changes

Run 'screen -ls' in a busy waiting loop to wait for the given regular expression object. 
If the object is not found within the default time frame, raise an exception.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided.
- [x] The description of changes is clear and encompassing.
- [x] No docs need to be updated as part of this PR.
- [x] Code-level documentation for touched code is included in this PR.
- [x] No API changes are included in this PR
- [x] The changes in this PR have no user impact
- [x] No new `unsafe` code has been added

### CODE REVIEW

File: tests/framework/utils.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (raduiliescu) [inline tests/framework/utils.py]: Maybe have been nicer to just have a function that would run a command, then verify with a regexp and use retry_call - https://pypi.org/project/retry/ for handling the retry part?

Comment (dianpopa) [reply tests/framework/utils.py]: I like @raduiliescu s suggestion. We already use `retry_call` in a bunch of places: https://github.com/firecracker-microvm/firecracker/blob/master/tests/framework/jailer.py#L201, https://github.com/firecracker-microvm/firecracker/blob/master/tests/framework/microvm.py#L357. We could stick to it across the CI framework.

Comment (gc-plp) [reply tests/framework/utils.py]: I updated the PR to use the reply module.

File: tests/framework/microvm.py [hunk 2]
```diff
@@ -333,10 +335,22 @@ def spawn(self):
...
-            out = run('screen -ls', shell=True, stdout=PIPE)\
-                .stdout.decode('utf-8')
-            screen_pid = re.search(r'([0-9]+)\.{}'.format(self._session_name),
-                                   out).group(1)
+            # Build a regex object to match (number).session_name
+            regex_object = re.compile(
+                r'([0-9]+)\.{}'.format(self._session_name))
+
+            # Check the output of 'screen -ls' until we get the PID or
+            # a timeout
```
Comment (iulianbarbu) [inline tests/framework/microvm.py]: What is this timeout referring to? It is the timeout specified in `pytest.ini`?

Comment (gc-plp) [reply tests/framework/microvm.py]: It refers to an exception thrown by the retry() call. I'll reword the comment.

File: tests/framework/microvm.py [hunk 3]
```diff
@@ -18,9 +18,11 @@
 from subprocess import run, PIPE
 
 from retry import retry
+from retry.api import retry_call
 
 import host_tools.memory as mem_tools
 import host_tools.network as net_tools
+import framework.utils as utils
```
Comment (aghecenco) [inline tests/framework/microvm.py]: nit: move this lower in the block of imports from `framework`

File: tests/framework/utils.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# SPDX-License-Identifier: Apache-2.0
+"""Generic utility functions that are used in the framework."""
+import subprocess
```
Comment (aghecenco) [inline tests/framework/utils.py]: nit: order importa alphabetically

Comment (gc-plp) [reply tests/framework/utils.py]: Will fix, thanks.

I think that these checks should be added in the linter. It should be doable to do it through one of the current linters.

**Comment by dianpopa** [comment]
gitlint complains:
```
1: T1 Title exceeds max length (51>50): "tests/licenses: add year 2020 as valid license year"
```

**Pull request merged**